nasa/cumulus
# CUMULUS-2478: Implement bucket map and /locate

**Summary:** Summary of changes

Addresses [CUMULUS-2478: Implement bucket map and /locate](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2478)

## Changes

* Implement [bucket map](https://github.com/asfadmin/thin-egress-app#bucket-mapping).  The `example/cumulus-tf/cumulus_distribution.tf` uses `example/cumulus-tf/cumulus_distribution/bucket_map.yaml.tmpl` to add bucket map file to `s3://_system_bucket_/_prefix_/cumulus_distribution/bucket_map.yaml`, and it can use `example/cumulus-tf/cumulus_distribution/bucket_map.yaml` instead if present
* Implement /locate endpoint
* Add 'Bearer token' support as an authorization method
* distribution api checks the file request against bucket map:
  - retrieves the bucket and key from file path
  - determine if the file request is public based on the `bucket map` rather than checking if the bucket type is 'public'
  - (EDL only) If the file request is not public, after authorization, it checks if the file is from private buckets on the bucket map. If it's a private bucket, check if the user has access based on the user groups setting.  This feature is currently only for EDL since Cognito doesn't support user_groups yet.

## Test instructions
* Test /locate endpoint: navigate to `_apiendpoint_/locate?bucket_name=_your-public-bucket_` should return 2 paths.  Because `example/cumulus-tf/cumulus_distribution/bucket_map.yaml.tmpl` adds two paths for each public bucket for testing purpose.
* Download file: (the s3 signed url can be found from JavaScript console)
  - Locate a file in _your-public-bucket_, navigate (without login) to `_api-url_/cumulus-daac/public-data/browse/_object-s3-key_`, and verify that we can download the file.  Check the s3 signed url, it should point to _your-public-bucket_ and _object-s3-key_ and `A-userid=A-userid=unauthenticated+user`
  - Try to download a file from protected bucket, `_api-url_/_protected-bucket_/_object-s3-key_`, it should direct to login and then download from s3 signed url.  The s3 signed url should have A-userid=yourloginuser.
  - While still logged in, try to download a public file.  The signed url should have A-userid=yourloginuser
  
* Test bearer token:
  - While still logged in, copy the accessToken from Developer JavaScript console or DynamoDb `_prefix_-DistributionAccessTokensTable`.
  - Go to / endpoint and logout from browser
  - Run `curl -verbose -L "https://api-url/protected-file-bucket-and-key" -o outputfile --header 'Authorization: Bearer ReplaceWithTheToken'`, and verify that file is downloaded. Check the screen output for s3 signed url, and it should have A-userid=yourusername

* Test bucket prefix:
`example/cumulus-tf/cumulus_distribution.tf` set bucketname_prefix to your bucket prefix, such as `cumulus-test-sandbox-`, and add `example/cumulus-tf/cumulus_distribution/bucket_map.yaml`.  You can copy one from `s3://_system_bucket_/_prefix_/cumulus_distribution/bucket_map.yaml` from previous deployment, and removes the bucket prefix from the bucketnames in the yaml file.
e.g. `jl-test-integration-protected: protected`, `PUBLIC_BUCKETS: public:`
After deployment, test /locate endpoint (e.g. locate?bucket_name=public), test public and non-public file downloads

* Download from Private Buckets (EDL only since cognito doesn't currently support user_groups):
  - add a  bucket to `example/cumulus-tf/cumulus_distribution/bucket_map.yaml` bucketmap. add bucket to  PRIVATE_BUCKETS section, and object prefix can be used, such as `data-bucket/pre-commission-data` where `pre-commission-data` is the object prefix.  See the `example example/cumulus-tf/cumulus_distribution/bucket_map.yaml.example`. Make sure your EDL user doesn't belong to the user groups listed.
  - download a file from the private bucket. The browse displays '403, This data is not currently available.'
  - update the user groups for the private bucket in the bucket map, so your EDL user belongs to one of the groups. Alternatively, you can update the accessToken record in `<prefix>-DistributionAccessTokensTable` to have tokenInfo.user_groups = ['one of the user groups for that private bucket'].  I have to manually update the record because I don't have any user_groups for my EDL account.  If you have one, you can add the group to the bucket map.
  - download a file from the private bucket

## PR Checklist
